gql/transport/phoenix_channel_websockets.py

```diff
@@ -0,0 +1,244 @@
+import asyncio
+import json
+from typing import Dict, Optional, Tuple
+
+from graphql import DocumentNode, ExecutionResult, print_ast
+from websockets.exceptions import ConnectionClosed
+
+from .exceptions import (
+    TransportProtocolError,
+    TransportQueryError,
+    TransportServerError,
+)
+from .websockets import WebsocketsTransport
+
+
+class PhoenixChannelWebsocketsTransport(WebsocketsTransport):
+    def __init__(
+        self, channel_name: str, heartbeat_interval: int = 30, *args, **kwargs
+    ) -> None:
+        self.channel_name = channel_name
+        self.heartbeat_interval = heartbeat_interval
+        self.subscription_ids_to_query_ids: Dict[str, int] = {}
+        super(PhoenixChannelWebsocketsTransport, self).__init__(*args, **kwargs)
+
+    async def _send_init_message_and_wait_ack(self) -> None:
+        """Join the specified channel and wait for the connection ACK.
+
+        If the answer is not a connection_ack message, we will return an Exception.
+        """
+
+        query_id = self.next_query_id
+        self.next_query_id += 1
+
+        init_message = json.dumps(
+            {
+                "topic": self.channel_name,
+                "event": "phx_join",
+                "payload": {},
+                "ref": query_id,
+            }
+        )
```

Comment:
My general though would be providing a [class typing](https://www.python.org/dev/peps/pep-0589/) for the json events as those are documented at [websocket-clients-and-phoenix-channels](http://graemehill.ca/websocket-clients-and-phoenix-channels/) made by [Graeme](https://elixirforum.com/t/specification-of-the-protocol-used-by-phoenix-channels/4192/7) on the linked Elixir Topic.

Unfortunately, this Python feature is available at Python 3.8+ as stated on the (library docs)(https://docs.python.org/3/library/typing.html#typing.TypedDict). In order to keep it backward compatible, the mypy suggest using [typing-extensions](https://mypy.readthedocs.io/en/stable/more_types.html#typeddict).

@leszekhanusz thoughts?

Reply:
IMO the typing hints are particularly interesting for prototype functions where they could be seen as a kind of documentation about the function arguments. I don't see the need to add them necessarily to all variables.

And in this case we don't really do that in the previous websockets.py code so I don't see why we should add it here.